annotation/src/main/java/org/cbioportal/genome_nexus/annotation/service/internal/VEPVariantAnnotationService.java

```diff
@@ -32,40 +32,118 @@
-        enrichmentService.enrichAnnotation(variantAnnotation);
+        VariantAnnotation variantAnnotation = variantAnnotationRepository.findOne(variant);
+        String annotationJSON = null;
+
+        if (variantAnnotation == null) {
+
+            try {
+                // get the annotation from the web service and save it to the DB
+                //variantAnnotation = variantAnnotationService.getAnnotation(variant);
+                //variantAnnotationRepository.save(variantAnnotation);
+
+                // get the raw annotation string from the web service
+                annotationJSON = this.getRawAnnotation(variant);
+
+                // construct a VariantAnnotation instance to return:
+                // this does not contain all the information obtained from the web service
+                // only the fields mapped to the VariantAnnotation model will be returned
+                variantAnnotation = this.mapAnnotationJson(variant, annotationJSON);
+
+                // save everything to the cache as a properly parsed JSON
+
+                try {
+                    variantAnnotationRepository.saveAnnotationJson(variant, annotationJSON);
+                }
+                catch (DataIntegrityViolationException e) {
+                    // in case of data integrity violation exception, do not bloat the logs
+                    // this is thrown when the annotationJSON can't be stored by mongo
+                    // due to the variant annotation key being too large to index
+                    LOG.info(e.getLocalizedMessage());
+                }
+            }
+            catch (HttpClientErrorException e) {
+                // in case of web service error, do not terminate the whole process.
+                // just copy the response body (error message) for this variant
+                variantAnnotation = new VariantAnnotation(variant, e.getResponseBodyAsString());
+            }
+            catch (IOException e) {
+                // in case of parse error, do not terminate the whole process.
+                // just send the raw annotationJSON to the client
+                variantAnnotation = new VariantAnnotation(variant, annotationJSON);
+            }
+        }
+
         return variantAnnotation;
     }
 
-    public String getRawAnnotation(String variant)
+    /**
+     * Maps the given raw annotation JSON string onto a VariantAnnotation instance.
+     *
+     * @param variant           variant key
+     * @param annotationJSON    raw annotation JSON string
+     * @return a VariantAnnotation instance
+     * @throws IOException
+     */
+    private VariantAnnotation mapAnnotationJson(String variant, String annotationJSON) throws IOException
+    {
+        // map annotation string onto VariantAnnotation instance
+        List<VariantAnnotation> list = Transformer.mapJsonToInstance(
+            annotationJSON, VariantAnnotation.class, new ExternalSourceObjectMapper());
```

Comment:
Would be better if those Object Mappers were singleton beans, instead of creating a new one every time with `new`.